File: 438.Find-All-Anagrams-in-a-String.java
// https://leetcode.com/problems/find-all-anagrams-in-a-string/
//
// algorithms
// Easy (36.94%)
// Total Accepted:    116,563
// Total Submissions: 315,507
// beats 100.0% of java submissions

class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        int sLen = s.length();
        int pLen = p.length();
        List<Integer> res = new ArrayList<>();
        
        if (pLen > sLen) {
            return res;
        }
                
        int pArr[] = new int[26];
        for (char ch : p.toCharArray()) {
            pArr[ch - 'a']++;
        }
        
        int sArr[] = new int[26];
        int i = 0;
        int j = 0;
        
        while (j < sLen) {
            int idx = s.charAt(j) - 'a'; 
            if (pArr[idx] == 0) {
                sArr = new int[26];
                i = j + 1;
            } else {
                sArr[idx]++;
                if (j - i + 1 == pLen) {
                    boolean isOk = true;
                    for (int k = 0; k < 26; k++) {
                        if (sArr[k] != pArr[k]) {
                            isOk = false;
                            break;
                        }
                    }
                    if (isOk) {
                        res.add(i);
                    }
                    sArr[s.charAt(i) - 'a']--;
                    i++;
                }
            }
            j++;
        }
        
        return res;
    }
}

File: 438.Find-All-Anagrams-in-a-String.py
# https://leetcode.com/problems/find-all-anagrams-in-a-string/
#
# algorithms
# Medium (37.8%)
# Total Accepted:    138,777
# Total Submissions: 365,240


class Solution(object):
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        len_s, len_p = len(s), len(p)
        if len_s < len_p:
            return []

        ch_p = [0] * 26
        for ch in p:
            ch_p[ord(ch) - 97] += 1

        ch_s = [0] * 26
        for i in xrange(len_p):
            ch_s[ord(s[i]) - 97] += 1

        res = []
        if self.equals(ch_p, ch_s):
            res += 0,

        for i in xrange(len_p, len_s):
            ch_s[ord(s[i]) - 97] += 1
            ch_s[ord(s[i - len_p]) - 97] -= 1
            if self.equals(ch_p, ch_s):
                res.append(i - len_p + 1)

        return res

    def equals(self, l1, l2):
        for i in xrange(26):
            if l1[i] != l2[i]:
                return False

        return True
